### DISCUSSION TITLE
storage createSignedUrl / createSignedUrls speedup

### DISCUSSION THREAD

**Discussion by gc-ft** [supabase/supabase#28692] [Feature Requests]
Hi,

the storage API's createSignedUrls interface is relatively "slow". For just a handful of Urls to sign it takes over half a second. In today's age of fast responding websites I would consider this slow.

After checking the code there seems to be a rather simple "speedup" fix, at least as an optional variable to the API calls:

For both a single object and for the multiple objects at once setup the system will verify that the objects in question exist and only then build a signed url. This process does not use anything from the findObject / findObjects calls, actually signObjectUrl depends on the findObject call to throw NoSuchKey error if objects are not found to stop the process since it is not caught in signObjectUrl.

I understand why this seems beneficial to do, however, if I am SURE that my objects exist (because I checked before) or similar reasons, why even check if they are there? Later on when the signed url is called to actually load the object it can still return a 404 if it does not exist. This seems faster behaviour than wasting time looking up the object when signing the url?

The only thing I can imagine is that somehow a signed url circumvents RLS of the objects and therefor the RLS needs to be checked before providing a signed url?

Fabian

**Comment by GaryAustin1** [comment]
RLS needs to be checked with the createSignedUrl call.  It generates the token that s3 will use to allow the request.  The actual use of the URL will not go back to the database to check RLS.

**Reply by gc-ft** [reply]
Ok, expected this somewhat.

Still, when getting signed URLs for objects using the service role, at that point verifying the object exists would not be needed?

This might be an edge case, but it will greatly increase signing performance in those cases and lower load on the superbase system?

**Reply by gc-ft** [reply]
Hi again,

I checked the code, this is actually not true!!

The /authenticated/ route to an object in supabase storage will use the jwt in the request to authenticate the db query that checks if the object exists, then loads it from s3 without anything authentication wise from the original request.

The /sign/ route will first validate the jwt based signature and then if successful will query the db asSuperuser() to lookup the object and then do the same exact s3 load as above.

The /public/ route will first verify the bucket in question is public (using asSuperuser()) and then lookup the object asSuperuser() and send it again via s3 load the same exact way.

This to me means still that the lookup of object when creating signed url is a waste of time, or at least should be optional?

**Reply by gc-ft** [reply]
Sorry, obviously missed the point of checking RLS when signing the url, that is true, currently this is when RLS is checked in these cases.

My previous comment stands though, when I know the user has access or when I would sign using the service role anyway, then there is no need to lookup the object and wasting db queries on it.

**Reply by GaryAustin1** [reply]
I was just commenting to you the reason they had to do the check in general to the database upfront because the s3 call did not do it.
